src/arithmetic/path_funcs/path_funcs.c

```diff
@@ -61,6 +64,149 @@ SIValue AR_TOPATH(SIValue *argv, int argc) {
 	return path;
 }
 
+// Routine for freeing a shortest path function's private data.
+void ShortestPath_Free(void *ctx_ptr) {
+	ShortestPathCtx *ctx = ctx_ptr;
+	if(ctx->reltypes) array_free(ctx->reltypes);
+	if(ctx->free_matrices) {
+		GrB_free(&ctx->R);
+		GrB_free(&ctx->TR);
+	}
+	rm_free(ctx);
+}
+
+// Routine for cloning a shortest path function's private data.
+void *ShortestPath_Clone(void *orig) {
+	ShortestPathCtx *ctx = orig;
+	// Allocate space for the clone
+	ShortestPathCtx *ctx_clone = rm_malloc(sizeof(ShortestPathCtx));
+	ctx_clone->minHops = ctx->minHops;
+	ctx_clone->maxHops = ctx->maxHops;
+	if(ctx->reltypes) array_clone(ctx_clone->reltypes, ctx->reltypes);
+	else ctx_clone->reltypes = NULL;
+	// Do not clone matrix data
+	ctx_clone->R = GrB_NULL;
+	ctx_clone->TR = GrB_NULL;
+	ctx_clone->free_matrices = false;
+
+	return ctx_clone;
+}
+
+SIValue AR_SHORTEST_PATH(SIValue *argv, int argc) {
+	if(SI_TYPE(argv[0]) == T_NULL) return SI_NullVal();
+	if(SI_TYPE(argv[1]) == T_NULL) return SI_NullVal();
+
+	Node *srcNode = argv[0].ptrval;
+	Node *destNode = argv[1].ptrval;
+	ShortestPathCtx *ctx = argv[2].ptrval;
+	GrB_Index src_id = ENTITY_GET_ID(srcNode);
+	GrB_Index dest_id = ENTITY_GET_ID(destNode);
+
+	GrB_Info res;
+	UNUSED(res);
+	Edge *edges = NULL;
+	GrB_Vector V = GrB_NULL;  // Vector of results
+	GrB_Vector PI = GrB_NULL; // Vector backtracking results to their parents.
+	GraphContext *gc = QueryCtx_GetGraphCtx();
+	uint reltype_count = (ctx->reltypes) ? array_len(ctx->reltypes) : 0;
+
+	/* The BFS algorithm uses a level of 1 to indicate the source node.
+	 * If this value is not zero (unlimited), increment it by 1
+	 * to make level 1 indicate the source's direct neighbors. */
+	int64_t max_level = (ctx->maxHops == EDGE_LENGTH_INF) ? 0 : ctx->maxHops + 1;
+
+	if(ctx->R == GrB_NULL) {
```

Comment:
This logic is performed once per query on the first invocation. The matrices are not cloned during `ShortestPath_Clone`; they're instead explicitly set to `GrB_NULL`.

This work is postponed to the first invocation and not included in the cloning routine because it would cause invalid assumptions in cached executions.